### ISSUE TITLE
Need help with Hyperparameters and Reward

### ISSUE THREAD

**Issue by cxovrika** [Unity-Technologies/ml-agents#331]
Hello

I'm trying to train agent in the game i wrote in unity. 
**GIF showing the game**: http://g.recordit.co/vykdH53OMt.gif (doing random stuff)

whole gameplay is in three buttons: **A/D** - corresponds to forcing your currently sticked left/right leg off the wall, **S** - forcing your currently floating leg to its corresponding wall (does nothing if both legs are sticked).

for training the agent has 4 possible actions (3 for A,D,S and one for doing nothing, sometimes it's better to wait, then press S)

I tried to train my agent with these **hyperparameters**: https://pastebin.com/RqcQA1PM

With the following reward function: 
reward = -0.002 by default,  also decreases by 0.05 if agent has chosen the action(A, D or S) that does nothing in that moment (trying to force a leg to a wall when its already sticked, or trying to force leg off the wall when its already floating), and at the end i add the Y velocity of the body(clamped [-0.1f, 0.4f], because falling down for a bit is ok, you cant avoid that). -1 for losing.

Also i have a 0.2 **Wait Time** in **Academy**, because when you play optimally the time between button pressings is nearly 0.3, there's no need to press button again instantly after sticking to a wall or forcing off the wall, so i think this must help the agent in training.

But my agent doesn't seem to train well, so i'm not sure if i'm doing it right, are my **Hyperparameters**, **Reward Function** and **Wait Time** good enough?

Thanks in advance for any help and suggestions!

**Comment by cxovrika** [comment]
Also i wanna say/ask about weird kinda bug thing. At the last agent step(when done = true) the reward was not going to the current episode's cumulative reward, instead setCumulativeReward() was called a bit late, after the agentReset() and before the first agent step of the next episode, so the last step's reward (which was -1) was counted as the next episode's first reward (if this was happening because i messed up something let me know please), so i just added a dummy step with reward of 0 at the end of every episode to solve this problem.

Has anyone seen the same behavior and is this way of fixing it ok?

**Comment by MarcoMeter** [comment]
Hi @cxovrika 
you described only your rewards, actions and environments. So what is the state space made of?

**Comment by cxovrika** [comment]
Hi @MarcoMeter 
it's almost the same as in crawler example:  current position, rotation and velocity for every limb and the body(middle head), acceleration of body, two booleans(0f / 1f) for each leg corresponding to if it is sticked to the wall at the moment, and two normalized angles for each leg/foot pair, and a bit more stuff from game mechanics, totally state size is 53. I think this information should be enough for training, totally

**Comment by jonboh** [comment]
Have you tried playing the game yourself? I've found myself wasting time with hyperparameter tuning having a buggy environment.


From your hyperparameter I would try to reduce the batch size to 32 or 64, and decrease the learning rate and epochs, to between 1e-5 and 1e-6, and the epochs to 3. I've found that bootstrapping the learning is easier with smaller batch sizes and smaller learning rate, that way it gets to quickly check a lot of possible strategies without commiting to much to anyone of them. Once the agent does more or less what I want I increase the batch size and slightly increase the learning rate aswell.

When does the agent get a positive reward? It is usually easier to train with positive rewards. One possibility would be to reward the increase of height from step to step.

**Comment by vincentpierre** [comment]
Several things that may be part of the problem: 
 * The cumulative reward in C# is not the same as the cumulative reward in python : The two are calculated independently. Do not use the C# cumulative reward as a source of truth to debug python.
 * The waitTime only works at inference time, in training, it is set to 0. Since your game uses physics, you cannot use waitTime.
 * Like @jonboh  said, try play the game yourself and see if it is feasible.

**Comment by cxovrika** [comment]
@vincentpierre  @jonboh  thanks for your help and suggestions!
I finally managed it to train properly ^^
Here's my agent if you wanna see: https://www.youtube.com/watch?v=jvB8niSO6uQ
